File: Solution O(n) space.java

import java.util.HashSet;
import java.util.Set;

public class Solution {

    public int[] findErrorNums(int[] input) {
        Set<Integer> unique = new HashSet<>();
        int expectedSum = input.length * (1 + input.length) / 2;
        int actualSum = 0;
        int duplicate = 0;
        for (int n : input) {
            if (!unique.add(n)) {
                duplicate = n;
            }
            actualSum += n;
        }
        return new int[]{duplicate, duplicate + expectedSum - actualSum};
    }
}

File: Solution O(n) space.js

/**
 * @param {number[]} input
 * @return {number[]}
 */
var findErrorNums = function (input) {
    const unique = new Set();
    let expectedSum = input.length * (1 + input.length) / 2;
    let actualSum = 0;
    let duplicate = 0;
    for (let n of input) {
        if (unique.has(n)) {
            duplicate = n;
        }
        unique.add(n);
        actualSum += n;
    }
    return [duplicate, duplicate + expectedSum - actualSum];
};
